Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+            }
+        }
+
+        /// <summary>
+        /// Gets the length of the text.
+        /// </summary>
+        public int Length => _text.Length;
+
+        protected bool IsShiftDown => GameService.Input.Keyboard.ActiveModifiers.HasFlag(ModifierKeys.Shift);
+        protected bool IsCtrlDown  => GameService.Input.Keyboard.ActiveModifiers.HasFlag(ModifierKeys.Ctrl);
+
+        protected bool _multiline;
+        protected bool _caretVisible;
+        protected bool _cursorMoved;
+
+        private TimeSpan _lastInvalidate;
+        private bool     _insertMode;
+        private bool     _suppressRedoStackReset;
+
+        private readonly UndoRedoStack _undoStack = new UndoRedoStack();
+        private readonly UndoRedoStack _redoStack = new UndoRedoStack();
+
+        private readonly Dictionary<Keys, KeyRepeatState> _keyRepeatStates;
+
+        public TextInputBase() {
+            _lastInvalidate  = DateTime.MinValue.TimeOfDay;
+            _keyRepeatStates = new Dictionary<Keys, KeyRepeatState>();
+
+            Input.Mouse.LeftMouseButtonReleased += OnGlobalMouseLeftMouseButtonReleased;
+            Input.Keyboard.KeyStateChanged      += OnGlobalKeyboardKeyStateChanged;
+        }
+
+        private void OnTextInput(string value) {
+            bool ctrlDown = this.IsCtrlDown;
+
+            foreach (char c in value) {
+                if (char.IsControl(c)) continue;
+                if (_font.GetCharacterRegion(c) == null) continue;
+
+                if (ctrlDown) {
+                    switch (c) {
+                        case 'c':
+                            HandleCopy();
+                            return;
+                        case 'x':
+                            HandleCut();
+                            return;
+                        case 'v':
+                            HandlePaste();
+                            return;
+                        case 'z':
+                            HandleUndo();
+                            return;
+                        case 'y':
+                            HandleRedo();
+                            return;
+                        case 'a':
+                            SelectAll();
+                            return;
+                    }
+                }
+
+                InputChar(c);
+            }
+        }
+
+        private void DeleteChars(int index, int length) {
+            if (length <= 0) return;
+
+            SetText(_text.Substring(0, index) + _text.Substring(index + length), true);
+        }
+
+        private string RemoveUnsupportedCharacters(string value) {
+            foreach (char c in value) {
+                if (_font.GetCharacterRegion(c) == null && !(_multiline && c == NEWLINE)) {
+                    return RemoveUnsupportedCharacters(value.Replace(c.ToString(), string.Empty));
+                }
+            }
+
+            return value;
+        }
+
+        private bool InsertChars(int index, string value, out int length) {
```

Comment:
We have currently a mix of abbreviation and non-abbreviation "char" and "character". Would be better to have one of them for consistency
You could also call the plural of characters a string

Reply:
There is currently:

- InsertChar
- InsertChars
- InputChar
- RemoveUnsupportedCharacters

I can rename the last to `RemoveUnsupportedChars` if you would prefer.

Reply:
I would prefer Characters because i personally dislike abbreviations (C/C++ nightmares) but i don't have any objective arguments for that matter so it's your decision.

Reply:
I've updated it to standardize to "Char(s)" simply to help match what is "upstream" in the Myra implementation in case we wish to pull future improvements to it into our own implementation since it'll be easier to follow the function names if they still match.